transports/webrtc/src/connection.rs

```diff
@@ -0,0 +1,271 @@
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+
+use std::{
+    sync::{Arc, Mutex as StdMutex},
+    task::{Context, Poll},
+};
+
+use crate::error::Error;
+pub(crate) use poll_data_channel::PollDataChannel;
+
+const MAX_DATA_CHANNELS_IN_FLIGHT: usize = 10;
+
+/// A WebRTC connection, wrapping [`RTCPeerConnection`] and implementing [`StreamMuxer`] trait.
+pub struct Connection {
+    /// `RTCPeerConnection` to the remote peer.
+    ///
+    /// Uses futures mutex because used in async code (see poll_outbound and poll_close).
+    peer_conn: Arc<FutMutex<RTCPeerConnection>>,
+
+    inner: StdMutex<ConnectionInner>,
+}
+
+struct ConnectionInner {
+    /// Channel onto which incoming data channels are put.
+    incoming_data_channels_rx: mpsc::Receiver<Arc<DetachedDataChannel>>,
+
+    /// Temporary read buffer's capacity (equal for all data channels).
+    /// See [`PollDataChannel`] `read_buf_cap`.
+    read_buf_cap: Option<usize>,
+
+    /// Future, which, once polled, will result in an outbound substream.
+    ///
+    /// NOTE: future might be waiting at one of the await points, and dropping the future will
+    /// abruptly interrupt the execution.
+    outbound_fut: Option<BoxFuture<'static, Result<Arc<DetachedDataChannel>, Error>>>,
+
+    /// Future, which, once polled, will result in closing the entire connection.
+    close_fut: Option<BoxFuture<'static, Result<(), Error>>>,
+}
+
+impl Connection {
+    /// Creates a new connection.
+    pub async fn new(rtc_conn: RTCPeerConnection) -> Self {
+        let (data_channel_tx, data_channel_rx) = mpsc::channel(MAX_DATA_CHANNELS_IN_FLIGHT);
+
+        Connection::register_incoming_data_channels_handler(&rtc_conn, data_channel_tx).await;
+
+        Self {
+            peer_conn: Arc::new(FutMutex::new(rtc_conn)),
+            inner: StdMutex::new(ConnectionInner {
+                incoming_data_channels_rx: data_channel_rx,
+                read_buf_cap: None,
+                outbound_fut: None,
+                close_fut: None,
+            }),
+        }
+    }
+
+    /// Set the capacity of a data channel's temporary read buffer (equal for all data channels; default: 8192).
+    pub fn set_data_channels_read_buf_capacity(&mut self, cap: usize) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.read_buf_cap = Some(cap);
+    }
+
+    /// Registers a handler for incoming data channels.
+    async fn register_incoming_data_channels_handler(
+        rtc_conn: &RTCPeerConnection,
+        tx: mpsc::Sender<Arc<DetachedDataChannel>>,
+    ) {
+        rtc_conn
+            .on_data_channel(Box::new(move |data_channel: Arc<RTCDataChannel>| {
+                debug!(
+                    "Incoming data channel '{}'-'{}'",
+                    data_channel.label(),
+                    data_channel.id()
+                );
+
+                let data_channel = data_channel.clone();
+                let mut tx = tx.clone();
```

Comment:
Note that for each call to `Clone::clone` a new slot is reserved in the capacity of the channel. See https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html

Thus, if I am not mistaken, this is just like using an `UnboundedSender`. Now I am not arguing in favor of using an unbounded channel. Quite the opposite. Any way we can get around this?

Reply:
Not sure I understand. I'm cloning a Sender, which is essentially an `Arc`. Where's "a new slot" is coming from?

Reply:
> The channel’s capacity is equal to buffer + num-senders. In other words, each sender gets a guaranteed slot in the channel capacity, and on top of that there are buffer “first come, first serve” slots available to all senders.

https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html

By cloning the `Sender` `buffer + num_senders` increases by `1`, i.e. adding "a new slot" to the overall capacity of the channel.

@melekes does that make sense? Happy to expand. I know this is not intuitive.

Reply:
Good catch :+1: Not sure yet about a way to resolve this.